src/udev.c

```diff
@@ -0,0 +1,590 @@
+  udev_find_more(dev, device, auto_assign);
+
+  if (auto_assign > 0)
+    policy_auto_assign_new_device(device);
+
+  return device;
+}
+
+static void
+udev_node_to_ids(const char *node, int *busid, int *devid)
+{
+  char *tmp;
+
+  /* USB devnodes look like "/dev/bus/usb/XXX/YYY",
+     XXX being the busid and YYY being the devid.
+     If other formats are ever encountered,
+     we may consider storing the devnode in device_t */
+
+  /* Instead of skipping 13 characters ("/dev/bus/usb/"),
+     just go to the first digit. */
+  while (*node != '\0' && (*node > '9' || *node < '0'))
+    node++;
+
+  /* strtol will stop at the next "/", and set tmp to its position */
+  *busid = strtol(node, &tmp, 10);
+
+  /* Instead of skipping 1 character ("/"), just go to the next digit. */
+  while (*tmp != '\0' && (*tmp < '0' || *tmp > '9' ))
+    tmp++;
+
+  *devid = strtol(tmp, NULL, 10);
+}
+
+/**
+ * Cleanup xenstore and delete a device after a udev removal event.
+ *
+ * @param dev Udev handle of the device
+ *
+ * @return 0 on success, 1 if nothing happened, -1 on failure
+ */
+int
+udev_del_device(struct udev_device *dev)
+{
+  const char *node;
+  const char *value;
+  int busnum;
+  int devnum;
+  usbinfo_t ui;
+  dominfo_t di;
+  device_t *device;
+  int ret;
+
+  /* Find the bus and device IDs */
+  node = udev_device_get_devnode(dev);
+  if (node == NULL)
+    return -1;
+  udev_node_to_ids(node, &busnum, &devnum);
+
+  /* Cleanup xenstore if the device was assigned to a VM */
+  device = device_lookup(busnum, devnum);
+  if (device == NULL) {
+    /* This happens if udev_maybe_add_device failed earlier, like on
+     * quick plug-unplug */
+    return 1;
+  }
+  if (device->vm != NULL) {
+    usbowls_build_usbinfo(busnum, devnum, device->vendorid, device->deviceid, &ui);
+    xenstore_get_dominfo(device->vm->domid, &di);
+    xenstore_destroy_usb(&di, &ui);
+  }
+
+  /* Delete the device from the global list */
+  ret = device_del(busnum, devnum);
+
+  return ret;
+}
+
+/**
+ * Enumerate all the udev USB devices that we care about,
+ * build nice model and vendor strings and add them to the list
+ */
+void
+udev_fill_devices(void)
+{
+  struct udev_enumerate *enumerate;
+  struct udev_list_entry *udev_device_list, *udev_device_entry;
+  struct udev_device *udev_device;
+  const char *path;
+  uint16_t vendor = 0;
+  uint16_t product = 0;
+
+  enumerate = udev_enumerate_new(udev_handle);
+  udev_enumerate_add_match_subsystem(enumerate, "usb");
+  /* Sysname must start with a digit */
+  udev_enumerate_add_match_sysname(enumerate, "[0-9]*");
+  udev_enumerate_scan_devices(enumerate);
+  udev_device_list = udev_enumerate_get_list_entry(enumerate);
+  udev_list_entry_foreach(udev_device_entry, udev_device_list) {
+    path = udev_list_entry_get_name(udev_device_entry);
+    udev_device = udev_device_new_from_syspath(udev_handle, path);
+    if (udev_maybe_add_device(udev_device, 0) == NULL)
+      udev_device_unref(udev_device);
+    else
```

Comment:
else-then-notta? perhaps put the "else" into the comment to get the point across without looking like an error?